cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentScheduler.java

```diff
@@ -0,0 +1,204 @@
+  private final NodeStatusProvider nodeStatusProvider;
+  private final ExecutorService agentWorkPool;
+  private final Integer parallelism;
+
+  private static Integer NOW = 0;
+  private static Integer INTERVAL = 30;
+  private static Integer TIMEOUT = 500;
+
+  private Map<String, AgentWorker> agents;
+  private volatile Integer runningAgents;
+
+  // This code assumes that every agent being run is in exactly either the WAITING or WORKING set.
+  private static final String WAITING_SET = "WAITZ";
+  private static final String WORKING_SET = "WORKZ";
+
+  ClusteredSortAgentScheduler(JedisSource jedisSource, NodeStatusProvider nodeStatusProvider, Integer parallelism) {
+    this.jedisSource = jedisSource;
+    this.nodeStatusProvider = nodeStatusProvider;
+    this.agents = new HashMap<>();
+    this.parallelism = parallelism;
+
+    if (this.parallelism == 0 || this.parallelism < -1) {
+      throw new IllegalArgumentException("Argument 'parallelism' must be positive, or -1 (for unlimited parallelism).");
+    }
+
+    this.runningAgents = 0;
+    this.agentWorkPool = Executors.newCachedThreadPool(new NamedThreadFactory(AgentWorker.class.getSimpleName()));
+    Executors.newSingleThreadScheduledExecutor(new NamedThreadFactory(ClusteredSortAgentScheduler.class.getSimpleName()))
+      .scheduleAtFixedRate(this, 0, 1, TimeUnit.SECONDS);
+  }
+
+  @Override
+  public void schedule(Agent agent, AgentExecution agentExecution, ExecutionInstrumentation executionInstrumentation) {
+    if (agent instanceof AgentSchedulerAware) {
+      ((AgentSchedulerAware)agent).setAgentScheduler(this);
+    }
+
+    try (Jedis jedis = jedisSource.getJedis()) {
+      // Regardless of the output, after this command it will be in one of the two sets.
+      jedis.eval(atomicAddScript(),
+          Arrays.asList(new String[]{WAITING_SET, WORKING_SET}),
+          Collections.singletonList(agent.getAgentType()));
+    }
+    agents.put(agent.getAgentType(), new AgentWorker(agent, agentExecution, executionInstrumentation, this));
+  }
+
+  @Override
+  public void run() {
+    if (!nodeStatusProvider.isNodeEnabled()) {
+      return;
+    }
+    try {
+      saturatePool();
+    } catch (Throwable t) {
+      t.printStackTrace(System.err);
+    }
+  }
+
+  private static String score(long offset) {
+    // Precision will be lost in this cast once 2^53 seconds have elapsed since epoch (~ 285 Million years from 1970).
+    // I'm optimistic about Spinnaker's success, but not optimistic enough to see this is as a problem.
+    return String.format("%f", (double)(TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis()) + offset));
+  }
+
+  // If the agent isn't present in either the WAITING or WORKING sets, it's safe to add. If it's present in either,
+  // it's being worked on or was recently run, so leave it be.
+  private static String atomicAddScript() {
+    return System.out.format("if redis.call('zrangebyscore', KEYS[1], ARGS[1]) != nil then\n" +
```

Comment:
it is more efficient in that 1) less bits over the wire (only passing the hash instead of the full script), and 2) the script is compiled once into redis's LUA memory and just invoked repeatedly
doing it this way where the actual executed script changes each time might end up creating a bunch of garbage in redis of previous scripts that will never execute again. at the very least its probably worth making the script itself static and parameterizing where it is String.formatting